### PR TITLE
Armbe8

### PR DESCRIPTION
**Pull request by jonwilson030981**
I have added support for ARMBE8. This architecture has little endian code, but big endian data. ARMBE32 has big endian code and data, but is considered legacy, accordingly I have not implemented it.

Also, TinyCC doesn't appear to support BE8, so that isn't implemented for now.

### CODE REVIEW

File: gum/arch-arm/gumarmbacktracer.c [hunk 1]
```diff
@@ -131,7 +131,8 @@ gum_arm_backtracer_generate (GumBacktracer * backtracer,
       {
         if (value % 4 == 0)
         {
-          const guint32 insn = *((guint32 *) GSIZE_TO_POINTER (value - 4));
+          const guint32 insn = GUINT32_FROM_LE(
```
Comment (oleavr) [inline gum/arch-arm/gumarmbacktracer.c]:
```suggestion
          const guint32 insn = GUINT32_FROM_LE (
```

File: gum/arch-arm/gumarmbacktracer.c [hunk 2]
```diff
@@ -131,7 +131,8 @@ gum_arm_backtracer_generate (GumBacktracer * backtracer,
...
+            *((guint32 *) GSIZE_TO_POINTER (value - 4)));
```
Comment (oleavr) [inline gum/arch-arm/gumarmbacktracer.c]:
```suggestion
              *((guint32 *) GSIZE_TO_POINTER (value - 4)));
```

Hanging indents should be four spaces, not two.

File: gum/arch-arm/gumarmbacktracer.c [hunk 3]
```diff
@@ -151,14 +152,14 @@ gum_arm_backtracer_generate (GumBacktracer * backtracer,
         else if ((value & 1) != 0)
         {
           const guint16 * insns_before = GSIZE_TO_POINTER (value - 1 - 2 - 2);
-          if ((insns_before[0] & 0xf800) == 0xf000 &&
-              (insns_before[1] & 0xe800) == 0xe800)
+          if ((GUINT16_FROM_LE(insns_before[0]) & 0xf800) == 0xf000 &&
```
Comment (oleavr) [inline gum/arch-arm/gumarmbacktracer.c]:
```suggestion
          if ((GUINT16_FROM_LE (insns_before[0]) & 0xf800) == 0xf000 &&
```

File: gum/arch-arm/gumarmbacktracer.c [hunk 4]
```diff
@@ -151,14 +152,14 @@ gum_arm_backtracer_generate (GumBacktracer * backtracer,
...
+              (GUINT16_FROM_LE(insns_before[1]) & 0xe800) == 0xe800)
```
Comment (oleavr) [inline gum/arch-arm/gumarmbacktracer.c]:
```suggestion
              (GUINT16_FROM_LE (insns_before[1]) & 0xe800) == 0xe800)
```

File: gum/arch-arm/gumarmbacktracer.c [hunk 5]
```diff
@@ -151,14 +152,14 @@ gum_arm_backtracer_generate (GumBacktracer * backtracer,
...
           {
             /* BL/BLX <imm11> */
             value--;
             valid = TRUE;
           }
-          else if ((insns_before[1] & 0xff80) == 0x4780)
+          else if ((GUINT16_FROM_LE(insns_before[1]) & 0xff80) == 0x4780)
```
Comment (oleavr) [inline gum/arch-arm/gumarmbacktracer.c]:
```suggestion
          else if ((GUINT16_FROM_LE (insns_before[1]) & 0xff80) == 0x4780)
```

File: gum/arch-x86/gumx86writer.c [hunk 6]
```diff
@@ -236,8 +236,11 @@ gum_x86_writer_flush (GumX86Writer * self)
         if (self->target_cpu == GUM_CPU_AMD64)
           *((guint64 *) (r->address - 8)) = GUINT64_TO_LE (target_pc);
         else
+#if G_BYTE_ORDER == G_LITTLE_ENDIAN
           *((guint32 *) (r->address - 4)) = GUINT32_TO_LE (target_pc);
-
+#else
+          *((guint32 *) (r->address - 4)) = GUINT32_TO_BE (target_pc);
+#endif
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: Let's add braces to both sides of this `if`, as this is a bit hard to read now.

File: tests/core/arch-arm/armrelocator.c [hunk 7]
```diff
@@ -260,13 +264,31 @@ branch_scenario_execute (BranchScenario * bs,
...
   g_assert_cmpuint (gum_arm_relocator_read_one (&fixture->rl, &insn), ==, 4);
   g_assert_cmpint (insn->id, ==, bs->instruction_id);
   g_assert_true (gum_arm_relocator_write_one (&fixture->rl));
   gum_arm_writer_flush (&fixture->aw);
-  g_assert_cmpint (memcmp (fixture->output, bs->expected_output,
-      bs->expected_output_length * sizeof (guint32)), ==, 0);
+  gboolean same_content = memcmp (fixture->output, bs->expected_output,
+      bs->expected_output_length * sizeof (guint32)) == 0;
+
+  gchar * diff = test_util_diff_binary (
```
Comment (oleavr) [inline tests/core/arch-arm/armrelocator.c]: Declaration after statement. Variables should be declared at the top of the nearest scope, in the order that they're used in. (Except for the value used to hold the return value, which should be first.)

File: gum/arch-x86/gumx86writer.c (hunk 6 above)
Comment (jonwilson030981) [reply gum/arch-x86/gumx86writer.c]: Ok. Think I got it.

File: tests/core/arch-arm/armrelocator.c [hunk 8]
```diff
@@ -260,13 +266,31 @@ branch_scenario_execute (BranchScenario * bs,
...
   g_assert_cmpint (insn->id, ==, bs->instruction_id);
   g_assert_true (gum_arm_relocator_write_one (&fixture->rl));
   gum_arm_writer_flush (&fixture->aw);
-  g_assert_cmpint (memcmp (fixture->output, bs->expected_output,
-      bs->expected_output_length * sizeof (guint32)), ==, 0);
+  same_content = memcmp (fixture->output, bs->expected_output,
+      bs->expected_output_length * sizeof (guint32)) == 0;
+
+  diff = test_util_diff_binary (
+      (guint8*)bs->expected_output,
```
Comment (oleavr) [inline tests/core/arch-arm/armrelocator.c]:
```suggestion
      (guint8 *) bs->expected_output,
```

File: tests/core/arch-arm/armrelocator.c [hunk 9]
```diff
@@ -260,13 +266,31 @@ branch_scenario_execute (BranchScenario * bs,
...
+      bs->expected_output_length * sizeof (guint32),
+      fixture->output,
+      bs->expected_output_length * sizeof (guint32));
+
+  if (!same_content)
+  {
+      g_print ("\n\nGenerated code is not equal to expected code:\n\n%s\n",
```
Comment (oleavr) [inline tests/core/arch-arm/armrelocator.c]:
```suggestion
    g_print ("\n\nGenerated code is not equal to expected code:\n\n%s\n",
```

File: tests/core/arch-arm/armrelocator.c [hunk 10]
```diff
@@ -260,13 +266,31 @@ branch_scenario_execute (BranchScenario * bs,
...
+          diff);
```
Comment (oleavr) [inline tests/core/arch-arm/armrelocator.c]:
```suggestion
        diff);
```

File: tests/core/arch-arm/armrelocator.c [hunk 11]
```diff
@@ -236,6 +236,8 @@ branch_scenario_execute (BranchScenario * bs,
 {
   gsize i;
   const cs_insn * insn = NULL;
+  gboolean same_content;
+  gchar * diff = NULL;
```
Comment (oleavr) [inline tests/core/arch-arm/armrelocator.c]:
```suggestion
  gchar * diff;
```

(Seems like a redundant initialization.)

**Comment by jonwilson030981** [comment]
Those changes all look fine. I think if we can just squash all into one that might be easiest. I can just about get by squashing adjacent commits, but trying to merge ones which aren't might be a bit too adventurous for me!

**Comment by oleavr** [comment]
Thanks, merged!

**Pull request merged**